### ISSUE TITLE
request: CUDA support

### ISSUE THREAD

**Issue by sunbubble** [universal-ctags/ctags#1220]
I've been using ctags with langmap c++ for CUDA ( *.cu and *.cuh filetypes), and it works just fine.
Is there a way to have this be default behavior?

$ ctags --langmap=c++:+.cu.cuh

**Comment by masatake** [comment]
@pragmaware, how do you think?

I got the same situation in https://github.com/universal-ctags/ctags/issues/1086 about jsx.
I accepted the patch (without agreement from @b4n, the maintainer of JavaScript parser.)
I'm still not sure this is good or not.

I don't know CUDA. As far as my quick reserach, it has some syntax level extension like `<<<...>>>`.
If a language has its own syntax, ctags should put the name of language to "language:" field.
So I don't want to say "ctags supports CUDA" with just adding " --langmap=c++:+.cu.cuh".

Ideally, CUDA subparser stacked on Cxx parser is implemented.
 keep in mind that your initial motivation of rewriting C and C++ parsers;
The original c.c was too complicated because it supports multiple `{' `}' syntax languages.

Just adding parserDefinition as a stub is one of approach.
```
diff --git a/parsers/cxx/cxx.c b/parsers/cxx/cxx.c
index a3e396d..e83c44a 100644
--- a/parsers/cxx/cxx.c
+++ b/parsers/cxx/cxx.c
@@ -135,3 +135,25 @@ parserDefinition * CppParser (void)
 
 	return def;
 }
+
+parserDefinition * CudaParser (void)
+{
+	static const char * const extensions [] =
+	{
+		"cu",
+		"cuh",
+		NULL
+	};
+	static parserDependency dependencies [] = {
+		{ DEPTYPE_KIND_OWNER, "C" },
+	};
+
+	static selectLanguage selectors[] = { selectByObjectiveCKeywords, NULL };
+
+	parserDefinition* def = parserNew("Cuda");
+
+	/* The same as C++. */
+	def->parser2 = cxxCppParserMain;
+	/* The same as C++. */
+	return def;
+}
```

At least we should require a test case as I did in #1086.

**Comment by pragmaware** [comment]
In theory CUDA is C with some additional keywords and builtin types. The only syntax difference that I'm aware of is the <<<>>> kernel calling syntax which probably can be ignored. 
The __global__ , __device__ host and similar keywords could be extracted and placed in the properties field, but this requires dedicated support for CUDA as language. This in turn would require a little bit of refactoring in the keyword management, adding another set of kinds and fields. Not very hard: I can do it.

So you tell me: do you want CUDA as a separate language with the support for the additional keywords or we keep it simple and just map .cu files to C?

**Comment by masatake** [comment]
> So you tell me: do you want CUDA 

me? or @sunbubble ?

If possible, implementing as separated language is better.  A tool can map to C after generating tags file.
Reporting different things in different terms is important job of ctags.

**Comment by pragmaware** [comment]
OK.

**Comment by sunbubble** [comment]
Whatever you think is best. Thank you.

**Comment by pragmaware** [comment]
Fixed via #1234

**Comment by sunbubble** [comment]
Thank you very much!
